hypothesis/client
# Get VitalSource chapter titles via alternate data source

Work around an issue where VitalSource's `getCurrentPage` API returns the wrong chapter titles [in some cases](https://github.com/hypothesis/client/issues/4986) by getting the data via an alternate `getTOC` API instead. This was recommended in discussion with one of their developers.

**Summary of changes:**

- Move CFI parsing utilities to `src/shared/` so they can be used by integrations in `src/annotator/integrations/`
- Change the logic for generating the `EPUBContentSelector` selector for VitalSource annotations, to get titles via the `getTOC` API if available, with a fallback to using the data from the existing source otherwise
- Implement `getTOC` API in our VS test pages

**Testing:**

Testing the issue as described on https://github.com/hypothesis/client/issues/4986 is difficult unless you happen to have the book mentioned in the issue in your VitalSource account. You can do some limited testing locally with the VS test pages.

1. Apply the following diff

```diff
diff --git a/dev-server/static/scripts/vitalsource-mosaic-book-element.js b/dev-server/static/scripts/vitalsource-mosaic-book-element.js
index 805073c63..42bb263fc 100644
--- a/dev-server/static/scripts/vitalsource-mosaic-book-element.js
+++ b/dev-server/static/scripts/vitalsource-mosaic-book-element.js
@@ -172,7 +172,7 @@ export class MosaicBookElement extends HTMLElement {
     // may be different. See https://github.com/hypothesis/client/issues/4986.
     const tocEntries = this.pageData.map(page => ({
       cfi: page.cfi,
-      title: page.chapterTitle,
+      title: page.chapterTitle + ' (from TOC)',
       path: page.absoluteURL,
       page: page.page,
     }));
```

2. Go to http://localhost:3000/document/vitalsource-epub and remove any existing annotations
3. Create some new annotations
4. The chapter titles in the sidebar should contain "(from TOC)"
